File: DONGUK/WEEK 6/2110.py
n, c = map(int,input().split())

dis = []
for _ in range(n):
    dis.append(int(input()))

dis.sort()

start, end = 1, (dis[-1] - dis[0] + 1)

res = 0
while start <= end:
    mid = (start + end) // 2  # 집 간 간격을 변수로 둔다

    cnt = 1
    tmp = dis[0]
    for i in range(1, n):
        if dis[i] - tmp >= mid:
            cnt += 1
            tmp = dis[i]

    
    if cnt >= c:
        start = mid + 1
        res = mid
    
    else:
        end = mid - 1

print(res)

File: KIJUNG/WEEK_06/2110.cpp
#include <iostream>
#include <algorithm>

#define FOR(i,a,b) for(int i=a;i<b;i++)

using namespace std;

int N, C;
int h[200000];
long long ans;

bool passible(long long n){
    long long cnt = 1;
    int tmp = h[0];
    FOR(i,1,N){
        if((h[i] - tmp)>=n) {
            tmp = h[i];
            cnt++;
            if(cnt >= C) {
                if(n > ans) ans = n;
                return true;
            }
        }
    }
    return false;
}

void solve(long long start, long long end){
    if(start > end) return;
    long long mid = (start + end)/2;
    if(passible(mid)) solve(mid+1,end);
    else solve(start, mid-1);
}

int main(void){
    cin >> N >> C;
    FOR(i,0,N){
        cin >> h[i];
    }
    sort(h,h+N);

    for(int i=1;;i*=2){
        if(!passible(i)) break;
    }

    solve(ans, ans*2);
    cout << ans;

    return 0;
}